airflow/providers/grafana/log/loki_task_handler.py

```diff
@@ -0,0 +1,242 @@
+        :param task_instance: task instance object
+        :type: task_instance: TaskInstance
+        """
+        tags = {}
+        task_labels = self._task_label(task_instance)
+        if self.labels:
+            tags.update(self.labels)
+            tags.update(task_labels)
+            return tags
+        return task_labels
+
+    def client(self, task_instance: TaskInstance) -> logging_loki.handlers.LokiHandler:
+        """
+        Returns Loki logger handler using connection and label to write logs
+        :param task_instance: task instance object
+        :type: task_instance: TaskInstance
+        """
+        return logging_loki.LokiHandler(
+            url=f"{self.get_conn.host}/loki/api/v1/push",
+            tags=self.get_label(task_instance),
+            version="1",
+            auth=(f"{self.get_conn.login}", f"{self.get_conn.password}"),
+        )
+
+    def set_context(self, task_instance: TaskInstance) -> None:
+        """
+        Provide task_instance context to airflow task handler.
+        If Loki is not available then will use Filehandler context
+        to avoid losing log data.
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        """
+        if self.is_loki_alive() == 200:
+            self.handler: logging_loki.handlers.LokiHandler = self.client(task_instance)
+            if self.formatter:
+                self.handler.setFormatter(self.formatter)
+                self.handler.setLevel(self.level)
+        else:
+            super().set_context(task_instance)  # Filehandler Context
+
+    def convert_epoch_time(self, timestamp: str) -> int:
+        """Convert timestamp to epoch"""
+        return int(time.mktime(time.strptime(timestamp[:19], self.timestamp_pattern)))
+
+    def query_time_range(
+        self,
+        task_instance: TaskInstance,
+    ) -> Tuple[int, int]:
+        """
+        Return task start and end time in epoch format for query.
+        If Task's try_number > 1 then it will use task execution time as start time.
+        :param ti: task instance object
+        :type: task_instance: TaskInstance
+        """
+        start_time, end_time, execution_time = (
+            str(task_instance.start_date.isoformat()),
+            task_instance.end_date,
+            str(task_instance.execution_date.isoformat()),
+        )
+        next_try = task_instance.next_try_number
+        start_epoch_time = self.convert_epoch_time(start_time)
+
+        if end_time and next_try < 3:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            return (start_epoch_time - 60, end_epoch_time + 60 * 2)
+        if end_time:
+            end_epoch_time = self.convert_epoch_time(end_time.isoformat())
+            execution_time_epoch_time = self.convert_epoch_time(execution_time)
+            return (execution_time_epoch_time - 60, end_epoch_time + 60 * 2)
+        return (start_epoch_time - 60, start_epoch_time + 60 * 60 * 24)
```

Comment:
Should end of range here be datetime.now instead? Otherwise it will not be possible to see logs for tasks in progress that have been running for over 24 hours. Maybe use `task_instance.execution_timeout` if set ?